File: src/main/java/uk/co/squadlist/web/controllers/SquadsController.java
package uk.co.squadlist.web.controllers;

import com.google.common.base.Splitter;
import com.google.common.collect.Lists;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.ModelAttribute;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.servlet.ModelAndView;
import uk.co.squadlist.client.swagger.ApiException;
import uk.co.squadlist.client.swagger.api.DefaultApi;
import uk.co.squadlist.model.swagger.Instance;
import uk.co.squadlist.model.swagger.Member;
import uk.co.squadlist.model.swagger.Squad;
import uk.co.squadlist.model.swagger.SquadSubmission;
import uk.co.squadlist.web.annotations.RequiresPermission;
import uk.co.squadlist.web.auth.LoggedInUserService;
import uk.co.squadlist.web.context.InstanceConfig;
import uk.co.squadlist.web.exceptions.SignedInMemberRequiredException;
import uk.co.squadlist.web.model.forms.SquadDetails;
import uk.co.squadlist.web.services.Permission;
import uk.co.squadlist.web.urls.UrlBuilder;
import uk.co.squadlist.web.views.NavItemsBuilder;
import uk.co.squadlist.web.views.ViewFactory;
import uk.co.squadlist.web.views.model.NavItem;

import javax.validation.Valid;
import java.io.IOException;
import java.net.URISyntaxException;
import java.util.List;

@Controller
public class SquadsController {

    private final static Logger log = LogManager.getLogger(SquadsController.class);

    private final static Splitter COMMA_SPLITTER = Splitter.on(",");

    private final UrlBuilder urlBuilder;
    private final ViewFactory viewFactory;
    private final LoggedInUserService loggedInUserService;
    private final NavItemsBuilder navItemsBuilder;
    private final InstanceConfig instanceConfig;
    private final DisplayMemberFactory displayMemberFactory;

    @Autowired
    public SquadsController(UrlBuilder urlBuilder,
                            ViewFactory viewFactory,
                            LoggedInUserService loggedInUserService,
                            NavItemsBuilder navItemsBuilder,
                            InstanceConfig instanceConfig,
                            DisplayMemberFactory displayMemberFactory
                            ) {
        this.urlBuilder = urlBuilder;
        this.viewFactory = viewFactory;
        this.loggedInUserService = loggedInUserService;
        this.navItemsBuilder = navItemsBuilder;
        this.instanceConfig = instanceConfig;
        this.displayMemberFactory = displayMemberFactory;
    }

    @RequestMapping(value = "/squad/new", method = RequestMethod.GET)
    public ModelAndView newSquad(@ModelAttribute("squadDetails") SquadDetails squadDetails) throws Exception {
        return renderNewSquadForm(new SquadDetails());
    }

    @RequestMapping(value = "/squad/new", method = RequestMethod.POST)
    public ModelAndView newSquadSubmit(@Valid @ModelAttribute("squadDetails") SquadDetails squadDetails, BindingResult result) throws SignedInMemberRequiredException, URISyntaxException, ApiException, IOException {
        DefaultApi swaggerApiClientForLoggedInUser = loggedInUserService.getSwaggerApiClientForLoggedInUser();
        Instance instance = swaggerApiClientForLoggedInUser.getInstance(instanceConfig.getInstance());

        if (result.hasErrors()) {
            return renderNewSquadForm(squadDetails);
        }

        try {
            SquadSubmission submission = new SquadSubmission().instance(instance).name(squadDetails.getName());
            swaggerApiClientForLoggedInUser.createSquad(submission);
            return viewFactory.redirectionTo(urlBuilder.adminUrl());

        } catch (ApiException e) {  // TODO more precise catch
            log.info("Invalid squad");
            result.rejectValue("name", null, "squad name is already in use");
            return renderNewSquadForm(squadDetails);
        }
    }

    @RequiresPermission(permission = Permission.VIEW_ADMIN_SCREEN)
    @RequestMapping(value = "/squad/{id}/delete", method = RequestMethod.GET)
    public ModelAndView deletePrompt(@PathVariable String id) throws SignedInMemberRequiredException, URISyntaxException, ApiException, IOException {
        DefaultApi swaggerApiClientForLoggedInUser = loggedInUserService.getSwaggerApiClientForLoggedInUser();
        Member loggedInUser = loggedInUserService.getLoggedInMember();
        Instance instance = swaggerApiClientForLoggedInUser.getInstance(instanceConfig.getInstance());
        List<Squad> squads = swaggerApiClientForLoggedInUser.getSquads(instance.getId());

        List<NavItem> navItems = navItemsBuilder.navItemsFor(loggedInUser, "admin", swaggerApiClientForLoggedInUser, instance, squads);

        final Squad squad = swaggerApiClientForLoggedInUser.getSquad(id);
        return viewFactory.getViewFor("deleteSquadPrompt", instance).
                addObject("title", "Delete squad").
                addObject("navItems", navItems).
                addObject("squad", squad);
    }

    @RequiresPermission(permission = Permission.VIEW_ADMIN_SCREEN)
    @RequestMapping(value = "/squad/{id}/delete", method = RequestMethod.POST)
    public ModelAndView delete(@PathVariable String id) throws SignedInMemberRequiredException, ApiException {
        DefaultApi swaggerApiClientForLoggedInUser = loggedInUserService.getSwaggerApiClientForLoggedInUser();

        final Squad squad = swaggerApiClientForLoggedInUser.getSquad(id);

        swaggerApiClientForLoggedInUser.deleteSquad(squad.getId());

        return viewFactory.redirectionTo(urlBuilder.adminUrl());
    }

    @RequestMapping(value = "/squad/{id}/edit", method = RequestMethod.GET)
    public ModelAndView editSquad(@PathVariable String id) throws SignedInMemberRequiredException, URISyntaxException, ApiException, IOException {
        DefaultApi swaggerApiClientForLoggedInUser = loggedInUserService.getSwaggerApiClientForLoggedInUser();
        Instance instance = swaggerApiClientForLoggedInUser.getInstance(instanceConfig.getInstance());
        Member loggedInUser = loggedInUserService.getLoggedInMember();

        final Squad squad = swaggerApiClientForLoggedInUser.getSquad(id);

        final SquadDetails squadDetails = new SquadDetails();
        squadDetails.setName(squad.getName());

        return renderEditSquadForm(loggedInUser, squad, squadDetails, swaggerApiClientForLoggedInUser, instance);
    }

    @RequestMapping(value = "/squad/{id}/edit", method = RequestMethod.POST)
    public ModelAndView editSquadSubmit(@PathVariable String id, @Valid @ModelAttribute("squadDetails") SquadDetails squadDetails, BindingResult result) throws IOException, SignedInMemberRequiredException, URISyntaxException, ApiException {
        DefaultApi swaggerApiClientForLoggedInUser = loggedInUserService.getSwaggerApiClientForLoggedInUser();
        Instance instance = swaggerApiClientForLoggedInUser.getInstance(instanceConfig.getInstance());
        Member loggedInMember = loggedInUserService.getLoggedInMember();

        final Squad squad = swaggerApiClientForLoggedInUser.getSquad(id);
        if (result.hasErrors()) {
            return renderEditSquadForm(loggedInMember, squad, squadDetails, swaggerApiClientForLoggedInUser, instance);
        }

        squad.setName(squadDetails.getName());
        log.info("Updating squad: " + squad);

        try {
            swaggerApiClientForLoggedInUser.updateSquad(squad, squad.getId());

        } catch (ApiException e) {  // TODO more precise exception
            log.warn("Invalid squad");
            return renderEditSquadForm(loggedInMember, squad, squadDetails, swaggerApiClientForLoggedInUser, instance);
        }

        final List<String> updatedSquadMembers = Lists.newArrayList(COMMA_SPLITTER.split(squadDetails.getMembers()).iterator());
        log.info("Setting squad members to " + updatedSquadMembers.size() + " members: " + updatedSquadMembers);
        swaggerApiClientForLoggedInUser.setSquadMembers(updatedSquadMembers, squad.getId());

        return viewFactory.redirectionTo(urlBuilder.adminUrl());
    }

    private ModelAndView renderNewSquadForm(SquadDetails squadDetails) throws SignedInMemberRequiredException, URISyntaxException, ApiException, IOException {
        DefaultApi swaggerApiClientForLoggedInUser = loggedInUserService.getSwaggerApiClientForLoggedInUser();
        final Member loggedInUser = loggedInUserService.getLoggedInMember();

        Instance instance = swaggerApiClientForLoggedInUser.getInstance(instanceConfig.getInstance());
        List<Squad> squads = swaggerApiClientForLoggedInUser.getSquads(instance.getId());

        List<NavItem> navItems = navItemsBuilder.navItemsFor(loggedInUser, "admin", swaggerApiClientForLoggedInUser, instance, squads);

        return viewFactory.getViewFor("newSquad", instance).
                addObject("title", "Add new squad").
                addObject("navItems", navItems).
                addObject("squadDetails", squadDetails);
    }

    private ModelAndView renderEditSquadForm(Member loggedInUser, Squad squad, SquadDetails squadDetails, DefaultApi api, Instance instance) throws SignedInMemberRequiredException, URISyntaxException, ApiException, IOException {
        final List<Member> squadMembers = api.getSquadMembers(squad.getId());
        DefaultApi swaggerApiClientForLoggedInUser = loggedInUserService.getSwaggerApiClientForLoggedInUser();
        List<Squad> squads = swaggerApiClientForLoggedInUser.getSquads(instance.getId());

        final List<Member> availableMembers = api.instancesInstanceMembersGet(instance.getId());
        availableMembers.removeAll(squadMembers);

        List<NavItem> navItems = navItemsBuilder.navItemsFor(loggedInUser, "admin", swaggerApiClientForLoggedInUser, instance, squads);

        return viewFactory.getViewFor("editSquad", instance).
                addObject("title", "Editing a squad").
                addObject("navItems", navItems).
                addObject("squad", squad).
                addObject("squadDetails", squadDetails).
                addObject("squadMembers", displayMemberFactory.toDisplayMembers(squadMembers, loggedInUser)).
                addObject("availableMembers", displayMemberFactory.toDisplayMembers(availableMembers, loggedInUser));
    }

}

File: src/main/kotlin/uk/co/squadlist/web/controllers/SquadsController.kt
package uk.co.squadlist.web.controllers

import com.google.common.base.Splitter
import com.google.common.collect.Lists
import org.apache.logging.log4j.LogManager
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Controller
import org.springframework.validation.BindingResult
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.ModelAttribute
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.servlet.ModelAndView
import uk.co.squadlist.client.swagger.ApiException
import uk.co.squadlist.client.swagger.api.DefaultApi
import uk.co.squadlist.model.swagger.Instance
import uk.co.squadlist.model.swagger.Member
import uk.co.squadlist.model.swagger.Squad
import uk.co.squadlist.model.swagger.SquadSubmission
import uk.co.squadlist.web.auth.LoggedInUserService
import uk.co.squadlist.web.context.InstanceConfig
import uk.co.squadlist.web.model.forms.SquadDetails
import uk.co.squadlist.web.services.PermissionsService
import uk.co.squadlist.web.urls.UrlBuilder
import uk.co.squadlist.web.views.NavItemsBuilder
import uk.co.squadlist.web.views.ViewFactory
import java.io.IOException
import java.net.URISyntaxException
import javax.validation.Valid

@Controller
class SquadsController @Autowired constructor(private val urlBuilder: UrlBuilder,
                                              private val viewFactory: ViewFactory,
                                              private val loggedInUserService: LoggedInUserService,
                                              private val navItemsBuilder: NavItemsBuilder,
                                              private val instanceConfig: InstanceConfig,
                                              private val displayMemberFactory: DisplayMemberFactory,
                                              permissionsService: PermissionsService,
                                              ): WithSignedInUser(instanceConfig, loggedInUserService, permissionsService) {

    private val log = LogManager.getLogger(SquadsController::class.java)

    private val COMMA_SPLITTER = Splitter.on(",")
    
    @GetMapping("/squad/new")
    fun newSquad(@ModelAttribute("squadDetails") squadDetails: SquadDetails?): ModelAndView {
        val renderAddSquadPage = { instance: Instance, loggedInMember: Member, swaggerApiClientForLoggedInUser: DefaultApi ->
            renderNewSquadForm(SquadDetails(), loggedInMember, instance, swaggerApiClientForLoggedInUser)
        }
        return withSignedInMemberWhoCanViewAdminScreen(renderAddSquadPage)
    }

    @PostMapping("/squad/new")
    @Throws(URISyntaxException::class, ApiException::class, IOException::class)
    fun newSquadSubmit(@Valid @ModelAttribute("squadDetails") squadDetails: SquadDetails?, result: BindingResult): ModelAndView {
        val handleAddSquad = { instance: Instance, loggedInMember: Member, swaggerApiClientForLoggedInUser: DefaultApi ->
            if (result.hasErrors()) {
                renderNewSquadForm(squadDetails, loggedInMember, instance, swaggerApiClientForLoggedInUser)
            } else try {
                val submission = SquadSubmission().instance(instance).name(squadDetails!!.name)
                swaggerApiClientForLoggedInUser.createSquad(submission)
                viewFactory.redirectionTo(urlBuilder.adminUrl())
            } catch (e: ApiException) {  // TODO more precise catch
                log.info("Invalid squad")
                result.rejectValue("name", null, "squad name is already in use")
                renderNewSquadForm(squadDetails, loggedInMember, instance, swaggerApiClientForLoggedInUser)
            }
        }
        return withSignedInMemberWhoCanViewAdminScreen(handleAddSquad)
    }

    @GetMapping("/squad/{id}/delete")
    @Throws(URISyntaxException::class, ApiException::class)
    fun deletePrompt(@PathVariable id: String): ModelAndView {
        val renderDeleteSquadPrompt = { instance: Instance, loggedInMember: Member, swaggerApiClientForLoggedInUser: DefaultApi ->
            val squads = swaggerApiClientForLoggedInUser.getSquads(instance.id)
            val navItems = navItemsBuilder.navItemsFor(loggedInMember, "admin", swaggerApiClientForLoggedInUser, instance, squads)
            val squad = swaggerApiClientForLoggedInUser.getSquad(id)
            viewFactory.getViewFor("deleteSquadPrompt", instance)
                    .addObject("title", "Delete squad")
                    .addObject("navItems", navItems)
                    .addObject("squad", squad)
        }
        return withSignedInMemberWhoCanViewAdminScreen(renderDeleteSquadPrompt)
    }

    @PostMapping("/squad/{id}/delete")
    @Throws(ApiException::class)
    fun delete(@PathVariable id: String): ModelAndView {
        val handleDeleteSquad = { instance: Instance, loggedInMember: Member, swaggerApiClientForLoggedInUser: DefaultApi ->
            val squad = swaggerApiClientForLoggedInUser.getSquad(id)
            swaggerApiClientForLoggedInUser.deleteSquad(squad.id)
            viewFactory.redirectionTo(urlBuilder.adminUrl())
        }
        return withSignedInMemberWhoCanViewAdminScreen(handleDeleteSquad)
    }

    @GetMapping("/squad/{id}/edit")
    @Throws(URISyntaxException::class, ApiException::class)
    fun editSquad(@PathVariable id: String): ModelAndView {
        val renderDeleteSquadPrompt = { instance: Instance, loggedInMember: Member, swaggerApiClientForLoggedInUser: DefaultApi ->
            val squad = swaggerApiClientForLoggedInUser.getSquad(id)
            val squadDetails = SquadDetails()
            squadDetails.name = squad.name
            renderEditSquadForm(loggedInMember, squad, squadDetails, instance, swaggerApiClientForLoggedInUser)
        }
        return withSignedInMemberWhoCanViewAdminScreen(renderDeleteSquadPrompt)
    }

    @PostMapping("/squad/{id}/edit")
    @Throws(URISyntaxException::class, ApiException::class)
    fun editSquadSubmit(@PathVariable id: String, @Valid @ModelAttribute("squadDetails") squadDetails: SquadDetails?, result: BindingResult): ModelAndView {
        val handleEditSquad = { instance: Instance, loggedInMember: Member, swaggerApiClientForLoggedInUser: DefaultApi ->

            val squad = swaggerApiClientForLoggedInUser.getSquad(id)
            if (result.hasErrors()) {
                renderEditSquadForm(loggedInMember, squad, squadDetails, instance, swaggerApiClientForLoggedInUser)
            }
            squad.name = squadDetails!!.name
            log.info("Updating squad: $squad")
            try {
                swaggerApiClientForLoggedInUser.updateSquad(squad, squad.id)
            } catch (e: ApiException) {  // TODO more precise exception
                log.warn("Invalid squad")
                renderEditSquadForm(loggedInMember, squad, squadDetails, instance, swaggerApiClientForLoggedInUser)
            }
            val updatedSquadMembers: List<String> = Lists.newArrayList(COMMA_SPLITTER.split(squadDetails.members).iterator())
            log.info("Setting squad members to " + updatedSquadMembers.size + " members: " + updatedSquadMembers)
            swaggerApiClientForLoggedInUser.setSquadMembers(updatedSquadMembers, squad.id)
            viewFactory.redirectionTo(urlBuilder.adminUrl())
        }
        return withSignedInMemberWhoCanViewAdminScreen(handleEditSquad)
    }

    @Throws(URISyntaxException::class, ApiException::class)
    private fun renderNewSquadForm(squadDetails: SquadDetails?, loggedInUser: Member, instance: Instance, swaggerApiClientForLoggedInUser: DefaultApi): ModelAndView {
        val squads = swaggerApiClientForLoggedInUser.getSquads(instance.id)
        val navItems = navItemsBuilder.navItemsFor(loggedInUser, "admin", swaggerApiClientForLoggedInUser, instance, squads)
        return viewFactory.getViewFor("newSquad", instance).addObject("title", "Add new squad").addObject("navItems", navItems).addObject("squadDetails", squadDetails)
    }

    @Throws(URISyntaxException::class, ApiException::class)
    private fun renderEditSquadForm(loggedInUser: Member, squad: Squad, squadDetails: SquadDetails?, instance: Instance, swaggerApiClientForLoggedInUser: DefaultApi): ModelAndView {
        val squadMembers = swaggerApiClientForLoggedInUser.getSquadMembers(squad.id)
        val squads = swaggerApiClientForLoggedInUser.getSquads(instance.id)
        val availableMembers = swaggerApiClientForLoggedInUser.instancesInstanceMembersGet(instance.id)
        availableMembers.removeAll(squadMembers)
        val navItems = navItemsBuilder.navItemsFor(loggedInUser, "admin", swaggerApiClientForLoggedInUser, instance, squads)
        return viewFactory.getViewFor("editSquad", instance)
                .addObject("title", "Editing a squad")
                .addObject("navItems", navItems)
                .addObject("squad", squad)
                .addObject("squadDetails", squadDetails)
                .addObject("squadMembers", displayMemberFactory.toDisplayMembers(squadMembers, loggedInUser))
                .addObject("availableMembers", displayMemberFactory.toDisplayMembers(availableMembers, loggedInUser))
    }

}